### ISSUE TITLE
Remove volatile flag

### ISSUE THREAD

**Issue by beam2d** [pfnet/chainer#2013]
The volatile flag of Variable has been used for saving the cost of constructing a graph when backprop is not needed. This approach is enough when a user who configures the volatility can access to all the input of the forward computation, but it cannot be used otherwise. We need a different approach with which we can disable the graph construction without modifying all the forward computation itself.

The no_backprop_mode is recently added, and can be used as above if all input variables have volatile=AUTO. Based on my experiences, this is enough for any use cases of volatile mode and is much easier to use. This proposal is to remove the volatility flag and make Variable behave like the current volatile=AUTO mode by default, whose graph construction behavior can be configured by a thread-local flag (like the current no_backprop_mode).

**Comment by unnonouno** [comment]
I tried to remove volatile flag from variable and to use `config.enable_backprop` instead. But i feel we cannot detect mixed volatility with this method. How do you think about this problme?

```
with chainer.no_backprop_mode():
    x = func()
y = x + 1
```

**Comment by beam2d** [comment]
IMO, any computations done with backprop mode enabled should build a computational graph. So, in the above case, func does not build a graph, but y = x+1 builds. I think there will be less chance of messing up two opposite attempts, "I want to backprop this part" and "I only want to do inference of this part", if we do not use volatile flag.

**Comment by beam2d** [comment]
Therefore, I think we do not need to store `_volatile` anymore. I might be missing some important case, so let me know if you find one.

**Comment by chenzhekl** [comment]
If I would like to fine-tune part of a network, I need to set the `volatile` attribute of the input variable to `off` and `on` after the frozen part. How would that be possible if `volatile` flag were to be removed?

**Comment by beam2d** [comment]
Execute the frozen part in the no-backprop mode, and run the rest in the usual mode. For example, the following code for v1

```python
x.volatile = 'on'
h = frozen_part(x)
h.volatile = 'off'
y = finetune_part(h)
```

is written in v2 as

```python
with chainer.no_backprop_mode():
    h = frozen_part(x)
y = finetune_part(h)
```

or equivalently

```python
chainer.config['enable_backprop'] = False
try:
    h = frozen_part(x)
finally:
    chainer.config['enable_backprop'] = True
y = finetune_part(h)
```

**Comment by chenzhekl** [comment]
Thanks, the new version is definitely clearer.

On Mar 29, 2017 11:07, "Seiya Tokui" <notifications@github.com> wrote:

Execute the frozen part in the no-backprop mode, and run the rest in the
usual mode. For example, the following code for v1

x.volatile = 'on'
h = frozen_part(x)
h.volatile = 'off'
y = finetune_part(h)

is written in v2 as

with chainer.no_backprop_mode():
    h = frozen_part(x)
y = finetune_part(h)

or equivalently

chainer.config['enable_backprop'] = Falsetry:
    h = frozen_part(x)finally:
    chainer.config['enable_backprop'] = True
y = finetune_part(h)

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/pfnet/chainer/issues/2013#issuecomment-289960125>, or mute
the thread

**Comment by unnonouno** [comment]
Fixed via #2356
